utils/yoke/src/yokeable.rs

```diff
@@ -0,0 +1,248 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::borrow::{Cow, ToOwned};
+use std::mem;
+
+/// A [`Yokeable`] type is essentially one with a covariant lifetime parameter,
+/// matched to the parameter in the trait definition. The trait allows one to cast
+/// the covariant lifetime to and from `'static`.
+///
+/// While Rust does not yet have GAT syntax, for the purpose of this documentation
+/// we shall refer to "`Self` with a lifetime `'a`" with the syntax `Self<'a>`.
+///
+/// [`Yokeable`]  exposes ways to cast between `Self<'static>` and `Self<'a>` generically.
+/// This is useful for turning covariant lifetimes to _dynamic_ lifetimes, where `'static` is
+/// used as a way to "erase" the lifetime.
+///
+/// # Implementation safety
+///
+/// This trait is safe to implement on types with a _covariant_ lifetime parameter, i.e. one where
+/// [`Self::transform()`]'s body can simply be `{ self }`. This will occur when the lifetime
+/// parameter is used within references, but not in the arguments of function pointers or in mutable
+/// positions (either in `&mut` or via interior mutability)
+///
+/// This trait must be implemented on the `'static` version of such a type, e.g. one should
+/// implement `Yokeable<'a>` (for all `'a`) on `Cow<'static, T>`.
+///
+/// There are further constraints on implementation safety on individual methods.
+///
+/// # Implementation example
+///
+/// This crate will eventually expose a custom derive that makes it possible to implement this
+/// trait safely without much ceremony. Such a custom derive will include static checks
+/// for the covariance of each field. In the meantime, this trait can typically be implemented as
+/// follows:
+///
+/// ```rust
+/// # use yoke::Yokeable;
+/// # use std::borrow::Cow;
+/// # use std::mem;
+/// struct Bar<'a> {
+///     numbers: Cow<'a, [u8]>,
+///     string: Cow<'a, str>,
+///     owned: Vec<u8>,
+/// }
+///
+/// unsafe impl<'a> Yokeable<'a> for Bar<'static> {
+///     type Output = Bar<'a>;
+///     fn transform(&'a self) -> &'a Bar<'a> {
+///         // covariant lifetime cast, can be done safely
+///         self
+///     }
+///
+///     unsafe fn make(from: Bar<'a>) -> Self {
+///         // We're just doing mem::transmute() here, however Rust is
+///         // not smart enough to realize that Bar<'a> and Bar<'static> are of
+///         // the same size, so instead we use transmute_copy
+///
+///         // This assert will be optimized out, but is included for additional
+///         // peace of mind as we are using transmute_copy
+///         debug_assert!(mem::size_of::<Bar<'a>>() == mem::size_of::<Self>());
+///         let ret = mem::transmute_copy(&from);
+///         mem::forget(from);
```

Comment:
Not safe in general; this may accidentally materialize aliasing references and other such nasties.

I'd just write the union version of this instead:

```rust
// Sprinkle in ManuallyDrop as necessary
union Transmute<A, B> {
  a: A, b: B,
}
Transmute { a: from }.b
```

Reply:
I'm curious, why would this be the case?

(The unchecked transmute is a bunch of code and I'd rather not recommend folks use more code for this. I'm also happy to `ptr::read` instead, if that's safer)

Reply:
Basically the danger here is: suppose we have `x: T<'a>` such that `x` has a mutable reference or some other "must be unique at all times" value. When you write:

```rust
let x: T<"a>;
let y = transmute_copy(&x);
// Danger!
forget(x);
```

at thje commented line, the value is duplicated, if ever so briefly. Now, I don't imagine this to be a *practical* concern, but this is the sort of place where crossing your i's and dotting your t's is key. If you want you could instead write

```rust
let x: T<'a>;
let ptr = &x as *const _;
forget(x);
ptr.cast().read()
```

Reply:
Ah, I can do that